src/backend/distributed/commands/foreign_server.c

```diff
@@ -0,0 +1,307 @@
+#include "distributed/metadata_sync.h"
+#include "distributed/worker_transaction.h"
+#include "foreign/foreign.h"
+#include "nodes/primnodes.h"
+
+static Node * RecreateForeignServerStmt(Oid serverId);
+
+
+/*
+ * PreprocessCreateForeignServerStmt is called during the planning phase for
+ * CREATE SERVER.
+ */
+List *
+PreprocessCreateForeignServerStmt(Node *node, const char *queryString,
+								  ProcessUtilityContext processUtilityContext)
+{
+	if (!ShouldPropagate())
+	{
+		return NIL;
+	}
+
+	EnsureCoordinator();
+
+	char *sql = DeparseTreeNode(node);
+
+	/* to prevent recursion with mx we disable ddl propagation */
+	List *commands = list_make3(DISABLE_DDL_PROPAGATION,
+								(void *) sql,
+								ENABLE_DDL_PROPAGATION);
+
+	return NodeDDLTaskList(NON_COORDINATOR_NODES, commands);
+}
+
+
+/*
+ * PreprocessAlterForeignServerStmt is called during the planning phase for
+ * ALTER SERVER .. OPTIONS ..
+ */
+List *
+PreprocessAlterForeignServerStmt(Node *node, const char *queryString,
+								 ProcessUtilityContext processUtilityContext)
+{
+	AlterForeignServerStmt *stmt = castNode(AlterForeignServerStmt, node);
+	ForeignServer *server = GetForeignServerByName(stmt->servername, false);
+	ObjectAddress address = { 0 };
+	ObjectAddressSet(address, ForeignServerRelationId, server->serverid);
+
+	if (!ShouldPropagateObject(&address))
+	{
+		return NIL;
+	}
+
+	EnsureCoordinator();
+
+	char *sql = DeparseTreeNode(node);
+
+	/* to prevent recursion with mx we disable ddl propagation */
+	List *commands = list_make3(DISABLE_DDL_PROPAGATION,
+								(void *) sql,
+								ENABLE_DDL_PROPAGATION);
+
+	return NodeDDLTaskList(NON_COORDINATOR_NODES, commands);
+}
+
+
+/*
+ * PreprocessRenameForeignServerStmt is called during the planning phase for
+ * ALTER SERVER RENAME.
+ */
+List *
+PreprocessRenameForeignServerStmt(Node *node, const char *queryString,
+								  ProcessUtilityContext processUtilityContext)
+{
+	RenameStmt *stmt = castNode(RenameStmt, node);
+	Assert(stmt->renameType == OBJECT_FOREIGN_SERVER);
+
+	char *serverName = strVal(stmt->object);
+	ForeignServer *server = GetForeignServerByName(serverName, false);
+	ObjectAddress address = { 0 };
+	ObjectAddressSet(address, ForeignServerRelationId, server->serverid);
+
+	/* filter distributed servers */
+	if (!ShouldPropagateObject(&address))
+	{
+		return NIL;
+	}
+
+	EnsureCoordinator();
+
+	char *sql = DeparseTreeNode(node);
+
+	/* to prevent recursion with mx we disable ddl propagation */
+	List *commands = list_make3(DISABLE_DDL_PROPAGATION,
+								(void *) sql,
+								ENABLE_DDL_PROPAGATION);
+
+	return NodeDDLTaskList(NON_COORDINATOR_NODES, commands);
+}
+
+
+/*
+ * PreprocessAlterForeignServerOwnerStmt is called during the planning phase for
+ * ALTER SERVER .. OWNER TO.
+ */
+List *
+PreprocessAlterForeignServerOwnerStmt(Node *node, const char *queryString,
+									  ProcessUtilityContext processUtilityContext)
+{
+	AlterOwnerStmt *stmt = castNode(AlterOwnerStmt, node);
```

Comment:
One thing that is tricky in this PR is alter owner. What if the new owner role is not on the workers?
```SQL
alter server myserver2 OWNER TO my_role;
ERROR:  role "my_role" does not exist
CONTEXT:  while executing command on localhost:9702
Time: 52.760 ms
```

I think that might be fine on the community, but on the enterprise we probably need `EnsureDependencies` call to make sure that the necessary objects(e.g., roles) are on the workers. So, we should still add `EnsureDependenciesExistOnAllNodes` here.

And, on the enteprise specific tests, we should add something like 
```SQL
-- prevent creating the roles just yet, let ALTER SERVER do that
SET citus.enable_object_propagation TO off;
create role ... ;
SET citus.enable_object_propagation TO on;

ALTER SERVER .. SET OWNER ...;
.. check owner is set properly on the workers ..
```